ddtrace/tracer.py

```diff
@@ -174,24 +202,35 @@ def start_span(self, name, child_of=None, service=None, resource=None, span_type
                 span_type=span_type,
             )
 
-            span.set_tag(system.PID, getpid())
-
-            #    http://pypi.datadoghq.com/trace/docs/#distributed-tracing
-            parent_trace_id, parent_span_id = context._get_parent_span_ids()
-            if parent_trace_id:
-                span.trace_id = parent_trace_id
-
-            if parent_span_id:
-                span.parent_id = parent_span_id
-
-            self.sampler.sample(span)
+            span.sampled = self.sampler.sample(span)
```

Comment:
What is the point of having a common interface for all samplers if we care about the concrete implementation and do type checks outside of it?

I thought the original API made more sense before. If I'm a `RateSampler` I know how to annotate the span accordingly. Also, we could have instead a `PrioritySampler` that first delegates the sampling to to the `RateSampler` and if the span is sampled, it also calls `RateByServiceSampler` and annotates it.

That way, the `Tracer` calls only `sampler.sample(span)` and we don't break encapsulation.

`PrioritySampler` would looks like:
```py
class PrioritySampler(object):
    def sample(self, span):
        sampled = rate_sampler.sample(span)
        by_service_sample.sample(span) if sampled
        return sampled
```

Reply:
You're quite right. But note that `PrioritySampler` would not necessarily call a `RateSampler` as the signature default sampler can be overridden in the constructor and for that part, we *do not* do type checking (internally, we actually replace it by something else on occasions) so the priority sampler should still run "any sampler that has a simple sample() func" if it decided to do everything itself. Mostly, the more we make this private, the better, because probably we don't want much external code to rely on the current implementation.

Reply:
@ufoot 
Oh, I see! If I understood correctly, we can still have something like the code below in the `.configure` method, right?
(so we let the underlying sampler be injected in the constructor)

```py
sampler = PrioritySampler(sampler) if priority_sampling
```